README.md
# dropwizard-micrometer

[Dropwizard][dropwizard] bundle that enables your dropwizard application for exposition of [micrometer-like][micrometer] 
metrics (system, jvm and http requests) as a [prometheus][prometheus] endpoint. In addition, if you use JDBI to manage 
mapping of objects to database tables, you can utilize [dropwizad-micrometer-jdbi](#dropwizad-micrometer-jdbi) package to meter SQL request's latencies.

## Packages
- `dropwizad-micrometer-core`
    Dropwizard bundle that implements prometheus endpoint and exposes core system metrics and JVM metrics utilizing micrometer instrumentation.
    Optionally, it provides servlet filter to record HTTP requests latencies and statuses within dimensional 
    [prometheus histogram][prometheus histogram].

- `dropwizad-micrometer-jdbi`
    An additional module to record latencies of [JDBI][jdbi] queries within dimensional [prometheus histogram][prometheus histogram].

## Usage
You can find an example of usage in the [ExampleApplication][dropwizard example application].

Below are the steps explained in more detail specifically for each package.

### dropwizad-micrometer-core
This package provides a minimal setup, i.e. it instantiates `/prometheus` endpoint, adds system and JVM metrics utilizing 
micrometer instrumentation, and optionally you can set up servlet filter to record HTTP requests latencies/statuses.

#### Add dependency into your `pom.xml`
If you use `maven`, you can simply reference it in the `<dependenccies>` block as below. 
The latest version can be found in [Releases](https://github.com/MaksymDolgykh/dropwizard-micrometer/releases) or in the maven [repository][dropwizard-micrometer-core maven repo]

```xml
    <dependencies>
        ...
        ...
        <dependency>
            <groupId>io.github.maksymdolgykh.dropwizard</groupId>
            <artifactId>dropwizard-micrometer-core</artifactId>
            <version>2.0.5</version>
        </dependency>
        ...
        ...
    </dependencies>


```
#### Import DropwizardMicrometer classes in your Application class
```java
import io.github.maksymdolgykh.dropwizard.micrometer.MicrometerBundle;
import io.github.maksymdolgykh.dropwizard.micrometer.MicrometerHttpFilter;
import javax.servlet.FilterRegistration;
import javax.servlet.DispatcherType;
import java.util.EnumSet;
```


#### Add the bundle to your application

Add `MicrometerBundle` class to the bootstrapping phase of your Application class

```java
public class ExampleApplication extends Application<ExampleConfiguration> {
    //...
    //...

    @Override
    public void initialize(Bootstrap<ExampleConfiguration> bootstrap) {
        //...
        //...
        bootstrap.addBundle(new MicrometerBundle());
        //...
        //...
    }
    //...
    //...
}
```

You will also need to make your `ExampleConfiguration` class implement `MicrometerBundleConfiguration` 
in order to provide the bundle with the configuration:
```java
public class ExampleConfiguration extends Configuration implements MicrometerBundleConfiguration {
    @JsonProperty("prometheus")
    private PrometheusConfiguration prometheusConfiguration = new PrometheusConfiguration();

    @Override
    public PrometheusConfiguration getPrometheusConfiguration() {
        return prometheusConfiguration;
    }
    //...
    //...
}
```
Assuming, the above example, configuration element is `prometheus`, so you can use configuration block like this in a config file:
```yaml
prometheus:
  endpoint: "/prometheus"
```

This will expose `/prometheus` endpoint in admin connector, by default admin connector is exposed at port `8081` 
in dropwizard apps.

### Assign servlet filter to the environment

To leverage latency metrics per http endpoint you need to assign servlet filter to the environment 
in your Application class within `run` method
```java
public class ExampleApplication extends Application<ExampleConfiguration> {
    //...
    //...
    @Override
    public void run(ExampleConfiguration configuration, Environment environment) {
        //...
        //...
        FilterRegistration.Dynamic micrometerFilter = environment.servlets().addFilter("MicrometerHttpFilter", new MicrometerHttpFilter());
        micrometerFilter.addMappingForUrlPatterns(EnumSet.allOf(DispatcherType.class), true, "/*");
    }
}
```
This way all http requests will be metered and metrics will be recorded within `http_server_requests_seconds` [histogram][prometheus histogram] 
with the following labels:
- `method`, http method, i.e. `GET`, `POST` etc
- `status`, response status, i.e. `200`, `404`, `500` etc
- `uri`, request path


### dropwizad-micrometer-jdbi
If you use [JDBI][jdbi] to manage mapping of objects to database tables, you can utilize 
[dropwizad-micrometer-jdbi][dropwizard-micrometer-jdbi maven repo] package to meter SQL request's latencies. 
It depends on `dropwizad-micrometer-core`, so that to use `dropwizad-micrometer-jdbi` package
the core package should be already installed and bundle should be added to your dropwizard application
(see how to install it in [dropwizad-micrometer-core](#dropwizad-micrometer-core) section). 
Once `dropwizad-micrometer-core` is installed, the steps to install `dropwizad-micrometer-jdbi` package are:

#### Add dependency into your `pom.xml`
If you use `maven`, you can simply reference it in the `<dependenccies>` block as below. 
The latest version can be found on in the maven [repository][dropwizard-micrometer-jdbi maven repo]

```xml
 <dependencies>
     ...
     ...
     <dependency>
         <groupId>io.github.maksymdolgykh.dropwizard</groupId>
         <artifactId>dropwizard-micrometer-jdbi</artifactId>
         <version>2.0.5</version>
     </dependency>
     ...
     ...
 </dependencies>

 ```

#### Import MicrometerJdbiTimingCollector class in your Application class
 ```java
import io.github.maksymdolgykh.dropwizard.micrometer.MicrometerJdbiTimingCollector;
 ```

#### Set TimingColletor for Jdbi object
To use the class you just need to set `TimingColletor` for Jdbi object, where it should be done depends on how your 
Application is organized - it might be in the `run` method of your `Application` class, or you might have separate 
class to configure DAO.
```
database.setTimingCollector(new MicrometerJdbiTimingCollector());
```
With this setup all jdbi requests will be metered and metrics will be recorded within `jdbi_requests_seconds` [histogram][prometheus histogram]

## License
This project is licensed under the Apache License 2.0 ([LICENSE](./LICENSE.TXT))


[dropwizard]: https://www.dropwizard.io/en/latest/
[dropwizard example application]: https://github.com/MaksymDolgykh/dropwizard-example-app
[dropwizard-micrometer-core maven repo]: https://mvnrepository.com/artifact/io.github.maksymdolgykh.dropwizard/dropwizard-micrometer-core
[dropwizard-micrometer-jdbi maven repo]: https://mvnrepository.com/artifact/io.github.maksymdolgykh.dropwizard/dropwizard-micrometer-jdbi
[jdbi]: https://jdbi.org/
[micrometer]: http://micrometer.io/
[prometheus]: https://prometheus.io/
[prometheus histogram]: https://prometheus.io/docs/practices/histograms/


CHANGELOG.md
# Changelog

## [3.0.0](https://github.com/MaksymDolgykh/dropwizard-micrometer/compare/v2.0.9...v3.0.0) (2022-12-14)


### ⚠ BREAKING CHANGES

* application Configuration class has to implement MicrometerBundleConfiguration

### Features

* make bundle configurable ([cbb1f93](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/cbb1f93f39a3edc3724d25af23093f0715215440)), closes [#19](https://github.com/MaksymDolgykh/dropwizard-micrometer/issues/19)


### Documentation

* update doc to include configuration step required to setup bundle ([21380c0](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/21380c04e518263d2c3918d7c84b47ce3971ccad))


### Miscellaneous Chores

* **master:** release 2.0.10-SNAPSHOT ([ae931d6](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/ae931d65d3ec110b2bda47f6c2c023cf57d26ae4))

## [2.0.9](https://github.com/MaksymDolgykh/dropwizard-micrometer/compare/v2.0.8...v2.0.9) (2022-11-28)


### Miscellaneous Chores

* additional micrometer binders ([34adce4](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/34adce428d1430769a8314f4a31e1ce17dd367a7))
* **master:** release 2.0.9-SNAPSHOT ([09859ce](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/09859ce1b5ac8bf76b52b6e0b3edeadd76c4eb0d))
* upgrade core dependencies for security patching (dropwizard, prometheus, micrometer) ([c25d74b](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/c25d74bd5ea481d3bbcf61210d1d5d5872f76454))

## [2.0.8](https://github.com/MaksymDolgykh/dropwizard-micrometer/compare/v2.0.7...v2.0.8) (2022-11-11)


### Miscellaneous Chores

* **master:** release 2.0.8-SNAPSHOT ([2927ddc](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/2927ddc7e045f7a0e8d5b2b20d2d779f292e12bc))

## [2.0.7](https://github.com/MaksymDolgykh/dropwizard-micrometer/compare/v2.0.6...v2.0.7) (2022-11-11)


### Miscellaneous Chores

* **master:** release 2.0.7-SNAPSHOT ([4168490](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/41684903490465dad831f77895a7683001de9cf4))

## [2.0.6](https://github.com/MaksymDolgykh/dropwizard-micrometer/compare/v2.0.5...v2.0.6) (2022-11-11)


### Continuous Integration

* add release workflow ([8d4a19d](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/8d4a19db0de8fb44f3af8ff71609ea6d7dc14f5f))
* add test-pr workflow ([5d49eda](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/5d49edaec145e554ddc50e08b280bee33e14fe7b))


### Documentation

* update docs ([1b8b1f0](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/1b8b1f0c0a6de3817c3560d70384b2f8931a8fb7))


### Miscellaneous Chores

* **master:** release 2.0.6-SNAPSHOT ([ee64a7b](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/ee64a7b40a3a246f5b54f2050e7b2236b0a574dc))
* release 2.0.6 ([89a5230](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/89a52301b75a78c750291feddc727cf6c8fe4223))

## [2.0.5](https://github.com/MaksymDolgykh/dropwizard-micrometer/compare/v1.0.0...v2.0.5) (2021-01-31)


### Features

* implement JDBI latency metrics and split to 2 packages (core and jdbi) ([bf2ce41](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/bf2ce4102ce3bbebafea4a22835c7f508ac500d3))

### Documentation

* update docs to reflect 2.0.x changes ([e5fd57a](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/e5fd57a86e7ccda99c9403827d55eff8460609fe))


## 1.0.0 (2020-09-30)


### Features

* add MicrometerBundle and servlet filter ([da224fd](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/da224fd3b5aecbcbfe9531c18ea4ec91bcf4cd3c))
* add integration with OSSRH ([1fe4541](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/1fe45413316dee2e6a074d0ffe0ba3467ab16efb))

### Tests

* add tests ([4260f9c](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/4260f9c8cd0726435ccc9a20e80341b45934eef5))

### Documentation

* add usage doc into readme ([f4c7751](https://github.com/MaksymDolgykh/dropwizard-micrometer/commit/f4c77513e628415a41c3a022629a3128df0de0c5))


core/src/test/java/io/github/maksymdolgykh/dropwizard/micrometer/MicrometerBundleTest.java
package io.github.maksymdolgykh.dropwizard.micrometer;


import javax.servlet.DispatcherType;
import javax.servlet.FilterRegistration;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.client.Client;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.dropwizard.Configuration;
import io.dropwizard.testing.ConfigOverride;
import io.dropwizard.testing.ResourceHelpers;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import io.dropwizard.Application;
import io.dropwizard.setup.Bootstrap;
import io.dropwizard.setup.Environment;
import io.dropwizard.testing.junit5.DropwizardAppExtension;
import io.dropwizard.testing.junit5.DropwizardExtensionsSupport;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.EnumSet;

import static org.junit.Assert.assertTrue;

@ExtendWith(DropwizardExtensionsSupport.class)
public class MicrometerBundleTest {

    private static int getFreePort() {
        ServerSocket socket = null;
        try {
            socket = new ServerSocket(0);
        } catch (IOException e) {
        }
        return socket.getLocalPort();
    }

    public static final DropwizardAppExtension<TestConfiguration> EXT = new DropwizardAppExtension<TestConfiguration>(TestApp.class,
            ResourceHelpers.resourceFilePath("testConfig.yaml"),
            ConfigOverride.config("server.applicationConnectors[0].port", String.valueOf(getFreePort())),
            ConfigOverride.config("server.adminConnectors[0].port", String.valueOf(getFreePort())));

    int adminPort = EXT.getAdminPort();
    int appPort = EXT.getLocalPort();
    Client client = EXT.client();

    PrometheusConfiguration prometheusConfiguration = EXT.getConfiguration().getPrometheusConfiguration();

    String ping = client.target("http://localhost:" + appPort + "/ping")
            .request().get(String.class);

    String scrape = client.target("http://localhost:"+ adminPort + prometheusConfiguration.getEndpoint())
            .request().get(String.class);
    

    @Test
    public void serverAcceptsAppConnection() throws IOException {
        Socket ableToConnect = new Socket("localhost", appPort);
        assertTrue(ableToConnect.isConnected());
        ableToConnect.close();
    }

    @Test
    public void testPingResource() {
        Assertions.assertThat(ping)
                .isNotNull()
                .matches("pong");
    }

    @Test
    public void serverAcceptsAdminConnection() throws IOException {
        Socket ableToConnect = new Socket("localhost", adminPort);
        assertTrue(ableToConnect.isConnected());
        ableToConnect.close();
    }

    @Test
    public void testScrappingEndpoint() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("# HELP", "# TYPE");
    }

    @Test
    public void testClassLoaderMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("jvm_classes_loaded", "jvm_classes_unloaded");
    }

    @Test
    public void testJvmMemoryMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("jvm_buffer_", "jvm_memory_");
    }

    @Test
    public void testJvmGcMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("jvm_gc_");
    }

    @Test
    public void testJvmThreadMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("jvm_threads_");
    }

    @Test
    public void testUptimeMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("process_uptime", "process_start_time");
    }

    @Test
    public void testProcessorMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("system_cpu_", "system_load_average");
    }

    @Test
    public void testLogbackMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("logback_events");
    }

    @Test
    public void testProcessFilesMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("process_files");
    }

    @Test
    public void testHttpLatencyMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("http_server_requests_seconds");
    }

    @Test
    public void testServletFilter() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .matches("(?s).*http_server_requests_seconds_bucket.*uri=\"/ping\".*");
    }

    public static class TestConfiguration extends Configuration implements MicrometerBundleConfiguration {

        @JsonProperty("prometheus")
        private PrometheusConfiguration prometheusConfiguration = new PrometheusConfiguration();

        @Override
        public PrometheusConfiguration getPrometheusConfiguration() {
            return prometheusConfiguration;
        }
    }


    @Path("/")
    public static class TestApp extends Application<TestConfiguration> {

        @Path("/ping")
        public static class PingResource {
            @GET
            public String ping() {
                return "pong";
            }
        }

        @Override
        public void initialize(Bootstrap<TestConfiguration> bootstrap){
            bootstrap.addBundle(new MicrometerBundle());
        }

        @Override
        public void run(TestConfiguration configuration, Environment environment) throws Exception {
            environment.jersey().register(new PingResource());

            FilterRegistration.Dynamic prometheusFilter = environment.servlets().addFilter("MicrometerHttpFilter", new MicrometerHttpFilter());
            prometheusFilter.addMappingForUrlPatterns(EnumSet.allOf(DispatcherType.class), true, "/*");

        }
    }
}


jdbi/src/test/java/io/github/maksymdolgykh/dropwizard/micrometer/MicrometerJdbiTimingCollectorTest.java
package io.github.maksymdolgykh.dropwizard.micrometer;

import javax.servlet.DispatcherType;
import javax.servlet.FilterRegistration;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.client.Client;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.dropwizard.testing.ConfigOverride;
import io.dropwizard.testing.ResourceHelpers;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import io.dropwizard.Application;
import io.dropwizard.Configuration;
import io.dropwizard.setup.Bootstrap;
import io.dropwizard.setup.Environment;
import org.jdbi.v3.core.Jdbi;
import io.dropwizard.testing.junit5.DropwizardAppExtension;
import io.dropwizard.testing.junit5.DropwizardExtensionsSupport;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.EnumSet;

import static org.junit.Assert.assertTrue;

@ExtendWith(DropwizardExtensionsSupport.class)
public class MicrometerJdbiTimingCollectorTest {

    private static int getFreePort() {
        ServerSocket socket = null;
        try {
            socket = new ServerSocket(0);
        } catch (IOException e) {
        }
        return socket.getLocalPort();
    }

    public static final DropwizardAppExtension<TestConfiguration> EXT = new DropwizardAppExtension<TestConfiguration>(TestApp.class,
            ResourceHelpers.resourceFilePath("testConfig.yaml"),
            ConfigOverride.config("server.applicationConnectors[0].port", String.valueOf(getFreePort())),
            ConfigOverride.config("server.adminConnectors[0].port", String.valueOf(getFreePort())));

    int adminPort = EXT.getAdminPort();
    int appPort = EXT.getLocalPort();
    Client client = EXT.client();

    PrometheusConfiguration prometheusConfiguration = EXT.getConfiguration().getPrometheusConfiguration();


    String ping = client.target("http://localhost:" + appPort + "/ping")
            .request().get(String.class);

    String scrape = client.target("http://localhost:"+ adminPort + prometheusConfiguration.getEndpoint())
            .request().get(String.class);
    

    @Test
    public void serverAcceptsAppConnection() throws IOException {
        Socket ableToConnect = new Socket("localhost", appPort);
        assertTrue(ableToConnect.isConnected());
        ableToConnect.close();
    }

    @Test
    public void serverAcceptsAdminConnection() throws IOException {
        Socket ableToConnect = new Socket("localhost", adminPort);
        assertTrue(ableToConnect.isConnected());
        ableToConnect.close();
    }

    @Test
    public void testScrappingEndpoint() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("# HELP", "# TYPE");
    }

    @Test
    public void testJDBIMetrics() {
        Assertions.assertThat(scrape)
                .isNotNull()
                .contains("jdbi_requests_seconds");
    }

    public static class TestConfiguration extends Configuration implements MicrometerBundleConfiguration {

        @JsonProperty("prometheus")
        private PrometheusConfiguration prometheus = new PrometheusConfiguration();

        @Override
        public PrometheusConfiguration getPrometheusConfiguration() {
            return prometheus;
        }
    }

    @Path("/")
    public static class TestApp extends Application<TestConfiguration> {

        @Path("/ping")
        public static class PingResource {
            @GET
            public String ping() {
                return "pong";
            }
        }

        @Override
        public void initialize(Bootstrap<TestConfiguration> bootstrap){
            bootstrap.addBundle(new MicrometerBundle());
        }

        @Override
        public void run(TestConfiguration configuration, Environment environment) throws Exception {
            Jdbi jdbi = Jdbi.create("jdbc:h2:mem:test");
            jdbi.setTimingCollector(new MicrometerJdbiTimingCollector());
            environment.jersey().register(new PingResource());

            FilterRegistration.Dynamic prometheusFilter = environment.servlets().addFilter("MicrometerHttpFilter", new MicrometerHttpFilter());
            prometheusFilter.addMappingForUrlPatterns(EnumSet.allOf(DispatcherType.class), true, "/*");
        }
    }
}


src/main/java/io/github/maksymdolgykh/dropwizard/micrometer/MicrometerHttpFilter.java
package io.github.maksymdolgykh.dropwizard.micrometer;

import javax.servlet.Filter;
import javax.servlet.FilterChain;
import javax.servlet.FilterConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpServletRequest;
import java.io.IOException;

public class MicrometerHttpFilter implements Filter {


    protected FilterConfig filterConfig;

    @Override
    public void init(FilterConfig filterConfig) throws ServletException {
        this.filterConfig = filterConfig;
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {

        // start the timer add let request to pass the filter
        long startTime = System.nanoTime();
        chain.doFilter(request, response);

        // record time after response
        double elapsed = (System.nanoTime() - startTime) / 1E9;
        String requestPath = ((HttpServletRequest) request).getPathInfo();

        String responseStatus = String.valueOf(((HttpServletResponse) response).getStatus());
        String requestMethod = ((HttpServletRequest) request).getMethod();
        MicrometerBundle.httpRequests.labels(requestMethod, responseStatus, requestPath).observe(elapsed);
    }

    @Override
    public void destroy() {
    }
}


jdbi/src/main/java/io/github/maksymdolgykh/dropwizard/micrometer/MicrometerJdbiTimingCollector.java
package io.github.maksymdolgykh.dropwizard.micrometer;

import org.jdbi.v3.core.statement.StatementContext;
import org.jdbi.v3.core.statement.TimingCollector;

import io.prometheus.client.Histogram;


public class MicrometerJdbiTimingCollector implements TimingCollector {

    private static final double[] jdbiLatencyBuckets = new double[]
            {.001, .002, .003, .004, .005, .007, .01, .015, .025, .05, .075,
                    .1, .15, .25, .35, .45, .5, .6, .7, .8, .9, 1,
                    1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9,
                    2, 2.5, 3, 3.5, 4, 4.5, 5,
                    5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5,
                    10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
                    20, 30, 40, 50, 60, 80, 100, 120};

    public static final Histogram jdbiRequests = Histogram.build()
            .name("jdbi_requests_seconds")
            .help("jdbi requests duration distribution")
            .labelNames("query")
            .buckets(jdbiLatencyBuckets)
            .register(MicrometerBundle.prometheusRegistry.getPrometheusRegistry());

    @Override
    public void collect(long elapsedTime, StatementContext ctx) {
        String sqlMethodName;

        try {
            sqlMethodName = ctx.getExtensionMethod().getMethod().getName();
        } catch (Exception e) {
            sqlMethodName = "undefined";
        }
        observePrometheusMetric(sqlMethodName, (elapsedTime / 1E9));
    }

    private void observePrometheusMetric (String sqlMethodName, double elapsedTime) {
        jdbiRequests.labels(sqlMethodName).observe(elapsedTime);
    }
}


core/src/main/java/io/github/maksymdolgykh/dropwizard/micrometer/MicrometerBundle.java
package io.github.maksymdolgykh.dropwizard.micrometer;

import io.dropwizard.ConfiguredBundle;
import io.dropwizard.setup.Bootstrap;
import io.dropwizard.setup.Environment;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.micrometer.core.instrument.binder.jvm.ClassLoaderMetrics;
import io.micrometer.core.instrument.binder.jvm.JvmGcMetrics;
import io.micrometer.core.instrument.binder.jvm.JvmHeapPressureMetrics;
import io.micrometer.core.instrument.binder.jvm.JvmMemoryMetrics;
import io.micrometer.core.instrument.binder.jvm.JvmThreadMetrics;
import io.micrometer.core.instrument.binder.system.FileDescriptorMetrics;
import io.micrometer.core.instrument.binder.system.ProcessorMetrics;
import io.micrometer.core.instrument.binder.system.UptimeMetrics;
import io.micrometer.core.instrument.binder.logging.LogbackMetrics;
import io.micrometer.prometheus.PrometheusConfig;
import io.micrometer.prometheus.PrometheusMeterRegistry;
import io.prometheus.client.exporter.MetricsServlet;
import io.prometheus.client.Histogram;

public class MicrometerBundle<T extends MicrometerBundleConfiguration> implements ConfiguredBundle<T> {
    private static final Logger LOGGER = LoggerFactory.getLogger(MicrometerBundle.class);

    private static final double[] httpLatencyBuckets = new double[]
    {.001, .002, .003, .004, .005, .007, .01, .015, .025, .05, .075,
            .1, .15, .25, .35, .45, .5, .6, .7, .8, .9, 1,
            1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9,
            2, 2.5, 3, 3.5, 4, 4.5, 5,
            5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5,
            10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
            20, 30, 40, 50, 60, 80, 100, 120};

    public static final PrometheusMeterRegistry prometheusRegistry = new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);

    public static final Histogram httpRequests = Histogram.build()
            .name("http_server_requests_seconds")
            .help("HTTP requests duration distribution")
            .labelNames("method", "status", "uri")
            .buckets(httpLatencyBuckets)
            .register(prometheusRegistry.getPrometheusRegistry());

    @Override
    public void run(T bundleConfiguration, Environment environment) throws Exception {
        PrometheusConfiguration prometheusConfiguration = bundleConfiguration.getPrometheusConfiguration();
        String endpoint = prometheusConfiguration.getEndpoint();

        // add system and jvm metrics
        new ClassLoaderMetrics().bindTo(prometheusRegistry);
        new JvmMemoryMetrics().bindTo(prometheusRegistry);
        new JvmGcMetrics().bindTo(prometheusRegistry);
        new JvmHeapPressureMetrics().bindTo(prometheusRegistry);
        new JvmThreadMetrics().bindTo(prometheusRegistry);
        new UptimeMetrics().bindTo(prometheusRegistry);
        new ProcessorMetrics().bindTo(prometheusRegistry);
        new LogbackMetrics().bindTo(prometheusRegistry);
        new FileDescriptorMetrics().bindTo(prometheusRegistry);

        MetricsServlet servlet = new MetricsServlet(prometheusRegistry.getPrometheusRegistry());

        LOGGER.info("Adding prometheus scraping endpoint as a servlet mapped to: {}", endpoint);
        environment.admin()
                .addServlet("prometheus", servlet)
                .addMapping(endpoint);
    }

    @Override
    public void initialize(Bootstrap<?> bootstrap) {
    }

    
}
